File: python/src/main/java/org/digma/intellij/plugin/psi/python/PythonCodeObjectsDiscovery.java
package org.digma.intellij.plugin.psi.python;

import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.PsiReference;
import com.intellij.psi.search.GlobalSearchScope;
import com.intellij.psi.search.searches.ReferencesSearch;
import com.intellij.psi.util.PsiTreeUtil;
import com.intellij.util.Query;
import com.jetbrains.python.psi.PyAssignmentStatement;
import com.jetbrains.python.psi.PyCallExpression;
import com.jetbrains.python.psi.PyExpression;
import com.jetbrains.python.psi.PyFile;
import com.jetbrains.python.psi.PyFunction;
import com.jetbrains.python.psi.PyReferenceExpression;
import com.jetbrains.python.psi.PyStringLiteralExpression;
import com.jetbrains.python.psi.PyTargetExpression;
import com.jetbrains.python.psi.stubs.PyFunctionNameIndex;
import org.apache.commons.lang3.time.StopWatch;
import org.digma.intellij.plugin.log.Log;
import org.digma.intellij.plugin.model.discovery.DocumentInfo;
import org.digma.intellij.plugin.model.discovery.MethodInfo;
import org.digma.intellij.plugin.model.discovery.SpanInfo;
import org.digma.intellij.plugin.psi.PsiUtils;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.TimeUnit;

public class PythonCodeObjectsDiscovery {

    private static final Logger LOGGER = Logger.getInstance(PythonCodeObjectsDiscovery.class);

    private static final String UNKNOWN_INST_LIBRARY = "UNKNOWN_INST_LIBRARY";
    private static final String UNKNOWN_SPAN_NAME = "UNKNOWN_SPAN_NAME";

    public static @NotNull DocumentInfo buildDocumentInfo(@NotNull Project project, @NotNull PyFile pyFile) {
        var stopWatch = StopWatch.createStarted();

        try {
            return buildDocumentInfoImpl(project, pyFile);
        } finally {
            stopWatch.stop();
            Log.log(LOGGER::debug, "buildDocumentInfo for {} took {} milliseconds", pyFile.getName(), stopWatch.getTime(TimeUnit.MILLISECONDS));
        }
    }


    private static @NotNull DocumentInfo buildDocumentInfoImpl(@NotNull Project project, @NotNull PyFile pyFile) {

        var fileUri = PsiUtils.psiFileToUri(pyFile);
        var methods = new HashMap<String, MethodInfo>();

        var functions = pyFile.getTopLevelFunctions();
        var classes = pyFile.getTopLevelClasses();

        classes.forEach(pyClass -> {
            for (PyFunction method : pyClass.getMethods()) {
                //function name should probably never be null but the interface is Nullable, so we need to check
                if (method.getName() != null) {
                    MethodInfo methodInfo = functionDiscovery(project, fileUri, method);
                    methods.put(methodInfo.getId(), methodInfo);
                }
            }
        });

        functions.forEach(pyFunction -> {
            //function name should probably never be null but the interface is Nullable, so we need to check
            if (pyFunction.getName() != null) {
                MethodInfo methodInfo = functionDiscovery(project, fileUri, pyFunction);
                methods.put(methodInfo.getId(), methodInfo);
            }
        });

        spanDiscovery(project, pyFile, fileUri, methods);
        return new DocumentInfo(fileUri, methods);
    }


    @NotNull
    private static MethodInfo functionDiscovery(@NotNull Project project, @NotNull String fileUri, @NotNull PyFunction pyFunction) {
        Objects.requireNonNull(pyFunction);
        Objects.requireNonNull(pyFunction.getName());

        var methodId = PythonLanguageUtils.createPythonMethodCodeObjectId(project, pyFunction);

        var name = pyFunction.getName();
        var className = pyFunction.getContainingClass() == null ? "" : pyFunction.getContainingClass().getName() == null ? "" : pyFunction.getContainingClass().getName();
        var namespace = pyFunction.getQualifiedName() == null ? "" : pyFunction.getQualifiedName();
        if (namespace.lastIndexOf(".py") > 0){
            namespace = namespace.substring(0, pyFunction.getQualifiedName().lastIndexOf(".py"));
        }

        var methodInfo = new MethodInfo(methodId, name, className, namespace, fileUri, pyFunction.getTextOffset());
        methodInfo.setAdditionalIdsProvider(new PythonAdditionalIdsProvider());
        return methodInfo;
    }


    private static void spanDiscovery(@NotNull Project project, @NotNull PyFile pyFile, String fileUri, @NotNull Map<String, MethodInfo> methodInfoMap) {
        spanDiscovery(project, pyFile, fileUri, Constants.OPENTELEMETRY_START_AS_CURRENT_SPAN_FUNC_NAME, methodInfoMap);
        spanDiscovery(project, pyFile, fileUri, Constants.OPENTELEMETRY_START_SPAN_FUNC_NAME, methodInfoMap);
    }

    private static void spanDiscovery(@NotNull Project project, @NotNull PyFile pyFile, String fileUri, @NotNull String tracerMethodName, @NotNull Map<String, MethodInfo> methodInfoMap) {

        var functions = PyFunctionNameIndex.find(tracerMethodName, project,GlobalSearchScope.allScope(project));

        //for some reason the search returns two identical functions, so just choose the first one.
        // I expect only one and don't know why there are two.
        PyFunction startSpanFunction = functions.stream().filter(pyFunction -> pyFunction.getContainingClass() != null &&
                Constants.OPENTELEMETRY_TRACER_FQN.equals(pyFunction.getContainingClass().getQualifiedName())).findFirst().orElse(null);

        if (startSpanFunction == null) {
            return;
        }

        Query<PsiReference> references = ReferencesSearch.search(startSpanFunction, GlobalSearchScope.fileScope(pyFile));
        references.forEach(psiReference -> {
            Log.log(LOGGER::debug, "found reference to {} function {}", tracerMethodName, psiReference.getElement().getText());
            var pyCallExpression = PsiTreeUtil.getParentOfType(psiReference.getElement(), PyCallExpression.class);
            if (pyCallExpression != null) {
                Log.log(LOGGER::debug, "call expression to {} function is {} ", tracerMethodName, pyCallExpression.getText());
                //span discovery may return more than one SpanInfo, but they all should belong
                // to the same method. it happens when a span has more than one id.
                List<SpanInfo> spanInfos = discoverSpanFromStartSpanMethodCallExpression(project, pyFile, pyCallExpression, fileUri);
                if (spanInfos.size() > 0) {
                    MethodInfo methodInfo = methodInfoMap.get(spanInfos.get(0).getContainingMethodId());
                    if (methodInfo != null) {
                        methodInfo.getSpans().addAll(spanInfos);
                    }
                }
            }
        });
    }


    /**
     * this method discovers a span from a call to one of the open telemetry start span methods.
     * the call expression is something like tracer.start_span("span name")
     */
    @NotNull
    public static List<SpanInfo> discoverSpanFromStartSpanMethodCallExpression(@NotNull Project project, @NotNull PsiFile pyFile, @NotNull PyCallExpression pyCallExpression, String fileUri) {

        var result = new ArrayList<SpanInfo>();
        var args = pyCallExpression.getArguments();
        var function = PsiTreeUtil.getParentOfType(pyCallExpression, PyFunction.class);

        //don't know how to compute span name if there are no arguments or if receiver is null or if the span is not inside a function block
        if (args.length >= 1 && function != null && pyCallExpression.getReceiver(null) != null) {

            var methodId = PythonLanguageUtils.createPythonMethodCodeObjectId(project, function);
            //span name is always the first argument to start_span or start_as_current_span
            var spanName = getSpanNameFromNameArgument(args[0]);
            //the receiver is the tracer object in this expression: tracer.start_span
            var receiver = pyCallExpression.getReceiver(null);
            //receiver should not be null, tested above
            var instLibrary = getInstLibraryFromReceiverExpression(Objects.requireNonNull(receiver));

            //in case the tracer name is the __name__ variable, create 3 spans with different instrumentation library names
            if (Constants.PYTHON_MODULE_NAME_VARIABLE.equals(instLibrary)) {

                //first instrumentation library name is "__main__"
                var spanId = PythonLanguageUtils.createSpanId("__main__", spanName);
                result.add(new SpanInfo(spanId, spanName, methodId, fileUri));

                //second instrumentation library is the relative path from project root or site-packages ,without the py extension.
                var relativePath = PythonLanguageUtils.getRelativePath(project,pyFile);
                if (relativePath.lastIndexOf(".py") >= 0){
                    relativePath = relativePath.substring(0, relativePath.lastIndexOf(".py"));
                }
                relativePath = relativePath.replace(File.separator, "."); //should work on linux and windows
                spanId = PythonLanguageUtils.createSpanId(relativePath, spanName);
                result.add(new SpanInfo(spanId, spanName, methodId, fileUri));

                //third instrumentation library is the relative path without its first segment if any
                if (relativePath.contains(".")) {
                    relativePath = relativePath.substring(relativePath.indexOf(".") + 1);
                    spanId = PythonLanguageUtils.createSpanId(relativePath, spanName);
                    result.add(new SpanInfo(spanId, spanName, methodId, fileUri));
                }

            } else {
                var spanId = PythonLanguageUtils.createSpanId(instLibrary, spanName);
                result.add(new SpanInfo(spanId, spanName, methodId, fileUri));
            }
        }

        return result;
    }


    //receiver expression is the tracer in tracer.start_span()
    @NotNull
    private static String getInstLibraryFromReceiverExpression(@NotNull PyExpression receiver) {

        if (receiver instanceof PyReferenceExpression referenceExpression) {
            return getInstLibraryFromReferenceExpression(referenceExpression);
        }else if(receiver instanceof PyCallExpression callExpression){
            return getInstLibraryFromCallExpression(callExpression);
        }

        return UNKNOWN_INST_LIBRARY;
    }


    @NotNull
    private static String getInstLibraryFromReferenceExpression(@NotNull PyReferenceExpression referenceExpression) {

        var psiElement = referenceExpression.getReference().resolve();
        if (psiElement != null) {
            PsiElement context = psiElement.getContext();
            if (context instanceof PyAssignmentStatement assignmentStatement) {
                return getInstLibraryFromAssignmentStatement(assignmentStatement);
            }
        }

        return UNKNOWN_INST_LIBRARY;
    }

    //assignmentStatement is the assignment in tracer = trace.get_tracer
    @NotNull
    private static String getInstLibraryFromAssignmentStatement(@NotNull PyAssignmentStatement assignmentStatement) {
        var assignedValue = assignmentStatement.getAssignedValue();
        if (assignedValue instanceof PyCallExpression callExpression) {
            return getInstLibraryFromCallExpression(callExpression);
        }

        return UNKNOWN_INST_LIBRARY;
    }

    //callExpression is the call to trace.get_tracer, we want the argument of get_tracer
    @NotNull
    private static String getInstLibraryFromCallExpression(@NotNull PyCallExpression callExpression) {
        var callee = callExpression.getCallee();
        if (callee != null && callee.getReference() != null) {
            var psiElement = callee.getReference().resolve();
            if (psiElement instanceof PyFunction pyFunction && Constants.OPENTELEMETRY_GET_TRACER_FUNC_NAME.equals(pyFunction.getName())) {
                var arguments = callExpression.getArguments();
                if (arguments.length >= 1) {
                    var arg = arguments[0];
                    return getInstLibraryFromGetTracerArgument(arg);
                }
            }
        }

        return UNKNOWN_INST_LIBRARY;
    }


    @NotNull
    private static String getInstLibraryFromGetTracerArgument(@NotNull PyExpression arg) {
        if (arg instanceof PyStringLiteralExpression stringLiteralExpression) {
            return getStringFromStringLiteralExpression(stringLiteralExpression);
        } else if (arg instanceof PyReferenceExpression referenceExpression) {
            return getInstLibraryFromArgumentReferenceExpression(referenceExpression);
        }

        return UNKNOWN_INST_LIBRARY;
    }


    @NotNull
    private static String getInstLibraryFromArgumentReferenceExpression(@NotNull PyReferenceExpression referenceExpression) {

        if (Constants.PYTHON_MODULE_NAME_VARIABLE.equals(referenceExpression.getText())) {
            return Constants.PYTHON_MODULE_NAME_VARIABLE;
        }

        var pyElement = referenceExpression.getReference().resolve();
        if (pyElement instanceof PyTargetExpression targetExpression){
            var assignedValue = targetExpression.findAssignedValue();
            if (assignedValue instanceof PyStringLiteralExpression stringLiteralExpression){
                return getStringFromStringLiteralExpression(stringLiteralExpression);
            }else if (assignedValue instanceof PyReferenceExpression pyReferenceExpressionAssignedValue){
                return getInstLibraryFromArgumentReferenceExpression(pyReferenceExpressionAssignedValue);
            }
        }

        return UNKNOWN_INST_LIBRARY;
    }


    @NotNull
    private static String getSpanNameFromNameArgument(@NotNull PyExpression pyExpression) {
        if (pyExpression instanceof PyStringLiteralExpression stringLiteralExpression) {
            return getStringFromStringLiteralExpression(stringLiteralExpression);
        }else if (pyExpression instanceof PyReferenceExpression referenceExpression){
            return getSpanNameFromReferenceExpression(referenceExpression);
        }

        return UNKNOWN_SPAN_NAME;
    }


    @NotNull
    private static String getSpanNameFromReferenceExpression(@NotNull PyReferenceExpression referenceExpression) {

        if (Constants.PYTHON_MODULE_NAME_VARIABLE.equals(referenceExpression.getText())) {
            return Constants.PYTHON_MODULE_NAME_VARIABLE;
        }

        var pyElement = referenceExpression.getReference().resolve();
        if (pyElement instanceof PyTargetExpression targetExpression){
            var assignedValue = targetExpression.findAssignedValue();
            if (assignedValue instanceof PyStringLiteralExpression stringLiteralExpression){
                return getStringFromStringLiteralExpression(stringLiteralExpression);
            }else if (assignedValue instanceof PyReferenceExpression pyReferenceExpressionAssignedValue){
                return getSpanNameFromReferenceExpression(pyReferenceExpressionAssignedValue);
            }
        }

        return UNKNOWN_SPAN_NAME;
    }


    @NotNull
    private static String getStringFromStringLiteralExpression(@NotNull PyStringLiteralExpression stringLiteralExpression) {
        return stringLiteralExpression.getStringValue();
    }


}

File: python/src/main/kotlin/org/digma/intellij/plugin/python/discovery/PythonCodeObjectsDiscovery.kt
@file:Suppress("UnstableApiUsage")

package org.digma.intellij.plugin.python.discovery

import com.intellij.lang.Language
import com.intellij.openapi.application.readAction
import com.intellij.openapi.application.smartReadAction
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiReference
import com.intellij.psi.search.GlobalSearchScope
import com.intellij.psi.search.ProjectScope
import com.intellij.psi.search.searches.ReferencesSearch
import com.intellij.psi.util.PsiTreeUtil
import com.intellij.refactoring.suggested.startOffset
import com.intellij.util.concurrency.annotations.RequiresReadLock
import com.jetbrains.python.psi.PyAssignmentStatement
import com.jetbrains.python.psi.PyCallExpression
import com.jetbrains.python.psi.PyExpression
import com.jetbrains.python.psi.PyFile
import com.jetbrains.python.psi.PyFunction
import com.jetbrains.python.psi.PyReferenceExpression
import com.jetbrains.python.psi.PyStringLiteralExpression
import com.jetbrains.python.psi.PyTargetExpression
import com.jetbrains.python.psi.stubs.PyFunctionNameIndex
import kotlinx.coroutines.ensureActive
import org.digma.intellij.plugin.common.GlobalSearchScopeProvider
import org.digma.intellij.plugin.model.discovery.DocumentInfo
import org.digma.intellij.plugin.model.discovery.MethodInfo
import org.digma.intellij.plugin.model.discovery.SpanInfo
import org.digma.intellij.plugin.python.PythonAdditionalIdsProvider
import org.digma.intellij.plugin.python.PythonMethodInfoDisplayNameProvider
import org.digma.intellij.plugin.python.createPythonMethodCodeObjectId
import org.digma.intellij.plugin.python.createSpanId
import org.digma.intellij.plugin.python.getPythonMethodNameFromMethodId
import org.digma.intellij.plugin.python.getRelativePath
import org.digma.intellij.plugin.python.isPythonMethodId
import java.io.File
import kotlin.coroutines.coroutineContext

const val PYTHON_MODULE_NAME_VARIABLE = "__name__"
const val OPENTELEMETRY_TRACER_FQN = "opentelemetry.trace.Tracer"
const val OPENTELEMETRY_START_AS_CURRENT_SPAN_FUNC_NAME = "start_as_current_span"
const val OPENTELEMETRY_START_SPAN_FUNC_NAME = "start_span"
const val OPENTELEMETRY_GET_TRACER_FUNC_NAME = "get_tracer"
const val SITE_PACKAGES_DIR_NAME = "site-packages"


const val UNKNOWN_INST_LIBRARY = "UNKNOWN_INST_LIBRARY"
const val UNKNOWN_SPAN_NAME = "UNKNOWN_SPAN_NAME"

suspend fun buildDocumentInfo(project: Project, pyFile: PyFile, fileUrl: String, language: Language): DocumentInfo? {

    val methodInfos = mutableMapOf<String, MethodInfo>()

    val topLevelFunctions = smartReadAction(project) { pyFile.topLevelFunctions.filter { it.name != null } }
    val topLevelClasses = smartReadAction(project) { pyFile.topLevelClasses }

    coroutineContext.ensureActive()
    topLevelClasses.forEach { pyClass ->
        val classFunctions = readAction { pyClass.methods.filter { it.name != null } }
        classFunctions.forEach { function ->
            coroutineContext.ensureActive()
            readAction {
                val methodInfo = functionDiscovery(project, fileUrl, function)
                methodInfos.put(methodInfo.id, methodInfo)
            }
        }
    }

    coroutineContext.ensureActive()
    topLevelFunctions.forEach { pyFunction ->
        coroutineContext.ensureActive()
        readAction {
            val methodInfo = functionDiscovery(project, fileUrl, pyFunction)
            methodInfos.put(methodInfo.id, methodInfo)
        }
    }

    coroutineContext.ensureActive()
    spanDiscovery(project, pyFile, fileUrl, methodInfos)
    return DocumentInfo(fileUrl, methodInfos, language.id)
}


@RequiresReadLock(generateAssertion = false)
fun functionDiscovery(project: Project, fileUrl: String, pyFunction: PyFunction): MethodInfo {
    val methodId = createPythonMethodCodeObjectId(project, pyFunction)
    val name = pyFunction.name ?: ""
    val containingClassName = pyFunction.containingClass?.name ?: ""
    val namespace = pyFunction.qualifiedName?.substringBefore(".py") ?: ""
    val methodInfo = MethodInfo(methodId, name, containingClassName, namespace, fileUrl)
    methodInfo.additionalIdsProvider = PythonAdditionalIdsProvider()
    methodInfo.displayNameProvider = PythonMethodInfoDisplayNameProvider()
    return methodInfo
}


suspend fun spanDiscovery(
    project: Project,
    pyFile: PyFile,
    fileUri: String,
    methods: MutableMap<String, MethodInfo>
) {
    spanDiscovery(project, pyFile, fileUri, OPENTELEMETRY_START_AS_CURRENT_SPAN_FUNC_NAME, methods)
    coroutineContext.ensureActive()
    spanDiscovery(project, pyFile, fileUri, OPENTELEMETRY_START_SPAN_FUNC_NAME, methods)
}


suspend fun spanDiscovery(
    project: Project,
    pyFile: PyFile,
    fileUri: String,
    tracerMethodName: String,
    methodInfoMap: MutableMap<String, MethodInfo>
) {
    val startSpanFunction = smartReadAction(project) {
        findFunctionByClassNameAndFunctionName(project, OPENTELEMETRY_TRACER_FQN, tracerMethodName) { ProjectScope.getAllScope(project) }
    } ?: return


    coroutineContext.ensureActive()

    val references = smartReadAction(project) {
        ReferencesSearch.search(startSpanFunction, GlobalSearchScope.fileScope(pyFile))
    }
    coroutineContext.ensureActive()

    references.forEach { psiReference: PsiReference ->
        val pyCallExpression = readAction {
            PsiTreeUtil.getParentOfType(psiReference.element, PyCallExpression::class.java)
        }
        pyCallExpression?.let { callExpression ->
            coroutineContext.ensureActive()
            val spanInfos = readAction {
                discoverSpanFromStartSpanMethodCallExpression(project, pyFile, callExpression, fileUri)
            }
            if (spanInfos.isNotEmpty()) {
                val methodInfo = methodInfoMap[spanInfos[0].containingMethodId]
                methodInfo?.spans?.addAll(spanInfos)
            }
        }
    }
}

@RequiresReadLock(generateAssertion = false)
fun discoverSpanFromStartSpanMethodCallExpression(
    project: Project,
    pyFile: PyFile,
    callExpression: PyCallExpression,
    fileUri: String
): List<SpanInfo> {
    val result = mutableListOf<SpanInfo>()
    val args = callExpression.arguments
    val function = PsiTreeUtil.getParentOfType(callExpression, PyFunction::class.java)

    val receiver = callExpression.getReceiver(null)
    if (args.isNotEmpty() && function != null && receiver != null) {
        val methodId = createPythonMethodCodeObjectId(project, function)
        val spanName = getSpanNameFromNameArgument(args[0])
        val instLibrary = getInstLibraryFromReceiverExpression(receiver)
        if (PYTHON_MODULE_NAME_VARIABLE == instLibrary) {
            var spanId = createSpanId("__main__", spanName)
            result.add(SpanInfo(spanId, spanName, methodId, fileUri, function.startOffset))

            var relativePath = getRelativePath(project, pyFile)
            if (relativePath.lastIndexOf(".py") >= 0) {
                relativePath = relativePath.substringBeforeLast(".py")
            }
            relativePath = relativePath.replace(File.separator, ".") //should work on linux and windows

            spanId = createSpanId(relativePath, spanName)
            result.add(SpanInfo(spanId, spanName, methodId, fileUri, function.startOffset))

            if (relativePath.contains(".")) {
                relativePath = relativePath.substringAfter(".")
                spanId = createSpanId(relativePath, spanName)
                result.add(SpanInfo(spanId, spanName, methodId, fileUri, function.startOffset))
            }
        } else {
            val spanId = createSpanId(instLibrary, spanName)
            result.add(SpanInfo(spanId, spanName, methodId, fileUri, function.startOffset))
        }
    }

    return result
}

@RequiresReadLock(generateAssertion = false)
fun getSpanNameFromNameArgument(pyExpression: PyExpression): String {
    if (pyExpression is PyStringLiteralExpression) {
        return getStringFromStringLiteralExpression(pyExpression)
    } else if (pyExpression is PyReferenceExpression) {
        return getSpanNameFromReferenceExpression(pyExpression)
    }

    return UNKNOWN_SPAN_NAME
}

@RequiresReadLock(generateAssertion = false)
fun getSpanNameFromReferenceExpression(referenceExpression: PyReferenceExpression): String {
    if (PYTHON_MODULE_NAME_VARIABLE == referenceExpression.text) {
        return PYTHON_MODULE_NAME_VARIABLE
    }

    val pyElement = referenceExpression.reference.resolve()
    if (pyElement is PyTargetExpression) {
        val assignedValue = pyElement.findAssignedValue()
        if (assignedValue is PyStringLiteralExpression) {
            return getStringFromStringLiteralExpression(assignedValue)
        } else if (assignedValue is PyReferenceExpression) {
            return getSpanNameFromReferenceExpression(assignedValue)
        }
    }

    return UNKNOWN_SPAN_NAME
}

@RequiresReadLock(generateAssertion = false)
fun getStringFromStringLiteralExpression(stringLiteralExpression: PyStringLiteralExpression): String {
    return stringLiteralExpression.stringValue
}


//receiver expression is the tracer in tracer.start_span()
@RequiresReadLock(generateAssertion = false)
fun getInstLibraryFromReceiverExpression(receiver: PyExpression): String {
    if (receiver is PyReferenceExpression) {
        return getInstLibraryFromReferenceExpression(receiver)
    } else if (receiver is PyCallExpression) {
        return getInstLibraryFromCallExpression(receiver)
    }

    return UNKNOWN_INST_LIBRARY
}

@RequiresReadLock(generateAssertion = false)
fun getInstLibraryFromReferenceExpression(referenceExpression: PyReferenceExpression): String {
    val psiElement = referenceExpression.reference.resolve()
    if (psiElement != null) {
        val context = psiElement.context
        if (context is PyAssignmentStatement) {
            return getInstLibraryFromAssignmentStatement(context)
        }
    }

    return UNKNOWN_INST_LIBRARY
}

@RequiresReadLock(generateAssertion = false)
fun getInstLibraryFromAssignmentStatement(assignmentStatement: PyAssignmentStatement): String {
    val assignedValue = assignmentStatement.assignedValue
    if (assignedValue is PyCallExpression) {
        return getInstLibraryFromCallExpression(assignedValue)
    }

    return UNKNOWN_INST_LIBRARY
}

@RequiresReadLock(generateAssertion = false)
fun getInstLibraryFromCallExpression(callExpression: PyCallExpression): String {
    val callee = callExpression.callee
    return callee?.reference?.let { reference ->
        val psiElement = reference.resolve()
        psiElement?.takeIf { it is PyFunction && it.name == OPENTELEMETRY_GET_TRACER_FUNC_NAME }?.let {
            val arguments = callExpression.arguments
            arguments.takeIf { it.isNotEmpty() }?.let { args ->
                val arg: PyExpression = args[0]
                getInstLibraryFromGetTracerArgument(arg)
            }
        }
    } ?: UNKNOWN_INST_LIBRARY
}

@RequiresReadLock(generateAssertion = false)
fun getInstLibraryFromGetTracerArgument(arg: PyExpression): String {
    if (arg is PyStringLiteralExpression) {
        return getStringFromStringLiteralExpression(arg)
    } else if (arg is PyReferenceExpression) {
        return getInstLibraryFromArgumentReferenceExpression(arg)
    }

    return UNKNOWN_INST_LIBRARY
}

@RequiresReadLock(generateAssertion = false)
fun getInstLibraryFromArgumentReferenceExpression(referenceExpression: PyReferenceExpression): String {
    if (PYTHON_MODULE_NAME_VARIABLE == referenceExpression.text) {
        return PYTHON_MODULE_NAME_VARIABLE
    }

    val pyElement = referenceExpression.reference.resolve()
    if (pyElement is PyTargetExpression) {
        val assignedValue = pyElement.findAssignedValue()
        if (assignedValue is PyStringLiteralExpression) {
            return getStringFromStringLiteralExpression(assignedValue)
        } else if (assignedValue is PyReferenceExpression) {
            return getInstLibraryFromArgumentReferenceExpression(assignedValue)
        }
    }

    return UNKNOWN_INST_LIBRARY
}

//the class index is not by qualified name, its short name index
//@RequiresReadLock(generateAssertion = false)
//fun findFunctionByClassNameAndFunctionName(
//    project: Project,
//    className: String,
//    functionName: String,
//    searchScopeProvider: GlobalSearchScopeProvider
//): PyFunction? {
//    val classes = PyClassNameIndex.find(className, project, searchScopeProvider.get())
//    for (pyClass in classes) {
//        val function = pyClass.findMethodByName(functionName, false, null)
//        if (function != null) {
//            return function
//        }
//    }
//    return null
//}

@RequiresReadLock(generateAssertion = false)
fun findFunctionByClassNameAndFunctionName(
    project: Project,
    className: String,
    functionName: String,
    searchScopeProvider: GlobalSearchScopeProvider
): PyFunction? {
    //the index will return a few functions, they have different containingClass, only one should match our search
    val functions = PyFunctionNameIndex.find(functionName, project, searchScopeProvider.get())
    return functions.firstOrNull { pyFunction: PyFunction -> pyFunction.isPhysical && pyFunction.containingClass?.qualifiedName == className }
}


@RequiresReadLock(generateAssertion = false)
fun findFunctionByMethodId(project: Project, methodId: String, searchScopeProvider: GlobalSearchScopeProvider): PyFunction? {
    return methodId.takeIf { isPythonMethodId(methodId) }?.let { id ->
        val methodName = getPythonMethodNameFromMethodId(id)
        PyFunctionNameIndex.find(methodName, project, searchScopeProvider.get())
            .firstOrNull { createPythonMethodCodeObjectId(project, it) == id }
    }


}